src/D2L.CodeStyle.Analyzers/Immutability/ImmutableDefinitionChecker.cs

```diff
@@ -219,28 +230,112 @@ ExpressionSyntax initializer
 				immutable = false;
 			}
 
-			var stuff = GetStuffToCheckForMember(
-				type,
-				typeSyntax,
-				initializer
-			);
+			// Check all initializations of the member for immutability
+			foreach( ExpressionSyntax initializer in initializers ) {
+				var stuff = GetStuffToCheckForMember(
+					type,
+					typeSyntax,
+					initializer
+				);
 
-			if ( stuff == null ) {
-				// null is a signal that there is nothing further that needs to
-				// be checked
-				return immutable;
-			}
+				if( stuff == null ) {
+					// Null is a signal that there is nothing further that needs to
+					// be checked
+					continue;
+				}
 
-			var (typeToCheck, checkKind, getLocation) = stuff.Value;
+				var (typeToCheck, checkKind, getLocation) = stuff.Value;
+
+				if( m_context.IsImmutable(
+					typeToCheck,
+					checkKind,
+					getLocation,
+					out var diagnostic
+				) ) {
+					continue;
+				}
 
-			if( !m_context.IsImmutable( typeToCheck, checkKind, getLocation, out var diagnostic ) ) {
 				diagnosticSink( diagnostic );
 				immutable = false;
 			}
 
 			return immutable;
 		}
 
+
+		/// <summary>
+		/// A readonly field or property may be re-initialized in constructors.
+		/// This method finds all initializations of the field/property.
+		/// </summary>
+		/// <param name="symbol">The field/property to search for</param>
+		/// <param name="initialInitializer">The initialization with the declaration</param>
+		/// <returns>An immutable list of initializations</returns>
+		private ImmutableArray<ExpressionSyntax> GetInitializers(
+			ISymbol symbol,
+			ExpressionSyntax initialInitializer
+		) {
+			List<ExpressionSyntax> initializers = new() {initialInitializer};
+
+			// If the containing symbol isn't a named type (class) then there
+			// will be no constructors
+			if( symbol.ContainingSymbol is not INamedTypeSymbol namedTypeSymbol ) {
+				return initializers.ToImmutableArray();
+			}
+
+			// Iterate through all constructors in the class
+			foreach( IMethodSymbol methodSymbol in namedTypeSymbol.Constructors ) {
+				// Make sure the syntax of the constructor exists
+				if( methodSymbol.DeclaringSyntaxReferences.Length == 0 ) {
+					continue;
+				}
+
+				// Retrieve the constructor syntax
+				var methodSyntax = methodSymbol.DeclaringSyntaxReferences
+				                               .Single()
+				                               .GetSyntax() as ConstructorDeclarationSyntax;
+
+				// Make sure the constructor actually contains a body
+				if( methodSyntax?.Body?.Statements == null ) {
+					continue;
+				}
+
+				// Retrieve the semantic model from the compilation
+				SemanticModel semanticModel = m_compilation.GetSemanticModel( methodSyntax.SyntaxTree );
+
+				// Iterate through every statement in the constructor
+				foreach( StatementSyntax statement in methodSyntax.Body.Statements ) {

```

Comment:
```suggestion
				foreach( var statement in methodSyntax.Body.Statements ) {
```

Can you go though the rest of the change and do this one? In the analyzer code we use `var` by default.

Reply:
Actually, this doesn't work:

```cs
public MyType() {
  if ( condition ) // x
  { // y
    m_foo = bar; // z
  }
}
```

The statement `z` isn't in `methodSyntax.Body.Statements`, it's a child statement of the block `y` which is attached to the (if) statement `x`.

There is an easier way to do this than nested loops though (you'd need recursion/a stack to fix this up):

```cs
var writes = namedTypeSymbol
    .Constructors
    .SelectMany( c => c.DescendentNodes() )
    .OfType<AssignmentExpressionSyntax>();
```

It's not as efficient as it could be, because there are cases where we would never find a descendent `AssignmentExpressionSyntax` (like inside of a `case` pattern, like you can't do `case (m_foo = bar):`) but it would be hard to actually implement that and possible fragile (the language can change).